extensions/eclipse-che-theia-plugin-ext/src/browser/plugin/che-plugin-manager.ts

```diff
@@ -0,0 +1,328 @@
+                if (registry === undefined) {
+                    instance.registryList.push({
+                        name: repoName,
+                        uri: uri
+                    });
+                }
+            });
+        }
+    }
+
+    private async initDefaults(): Promise<void> {
+        if (!this.defaultRegistry) {
+            this.defaultRegistry = await this.chePluginService.getDefaultRegistry();
+        }
+
+        if (!this.activeRegistry) {
+            this.activeRegistry = this.defaultRegistry;
+        }
+
+        if (!this.registryList) {
+            this.registryList = [this.defaultRegistry];
+
+            this.registryList.push({
+                name: 'Visual Studio Code plugins',
+                uri: 'https://marketplace.visualstudio.com/vscode'
+            });
+
+            await this.restoreRegistryList();
+        }
+
+        if (!this.installedPlugins) {
+            // Get list of plugins from workspace config
+            this.installedPlugins = await this.chePluginService.getWorkspacePlugins();
+        }
+    }
+
+    getDefaultRegistry(): ChePluginRegistry {
+        return this.defaultRegistry;
+    }
+
+    changeRegistry(registry: ChePluginRegistry): void {
+        this.activeRegistry = registry;
+        this.pluginRegistryChanged.fire(registry);
+    }
+
+    addRegistry(registry: ChePluginRegistry): void {
+        this.registryList.push(registry);
+
+        // Save list of custom repositories to preferences
+        const prefs = {};
+        this.registryList.forEach(r => {
+            if (r.name !== 'Default') {
+                prefs[r.name] = r.uri;
+            }
+        });
+
+        this.preferenceService.set('chePlugins.repositories', prefs, PreferenceScope.User);
+    }
+
+    removeRegistry(registry: ChePluginRegistry): void {
+        this.registryList = this.registryList.filter(r => r.uri !== registry.uri);
+    }
+
+    getRegistryList(): ChePluginRegistry[] {
+        return this.registryList;
+    }
+
+    /**
+     * Returns plugin list from active registry
+     */
+    async getPlugins(filter: string): Promise<ChePluginMetadata[]> {
+        await this.initDefaults();
+
+        this.availablePlugins = await this.chePluginService.getPlugins(this.activeRegistry, filter);
+        // The code will be use soon
+        // const deployedPlugins = await this.pluginFrontentService.getDeployedPlugins(filter);
+        // this.availablePlugins = this.availablePlugins.concat(deployedPlugins);
+        return this.availablePlugins;
+    }
+
+    isPluginInstalled(plugin: ChePluginMetadata): boolean {
+        return this.installedPlugins.indexOf(plugin.key) >= 0;
+    }
+
+    async install(plugin: ChePluginMetadata): Promise<boolean> {
+        try {
+            // add the plugin to workspace configuration
+            await this.chePluginService.addPlugin(plugin.key);
+            await this.delay(1000);
```

Comment:
Could you explain why do you need a delay here?

Reply:
Fixed. Thanks!